### DISCUSSION TITLE
Highlighting and manipulating parts of a raw block

### DISCUSSION THREAD

**Discussion by xbmbmx** [typst/typst#3057] [Q&A]
For some of my documents, I use `hex` content, often taken from the output of `hexdump -C <some-file>`.

Within this `hex` block, I require to mark some parts (bytes or sequences of bytes) to be able to add comments on them and reference them in the text later on. Similar to how wireshark highlights text in a dump.

In `LaTeX`, using the `lstlisting` package, this could be achieved like this:

```latex
\begin{lstlisting}[escapechar=@]
    AA @\colorbox{red}{BB}@ CC
\end{lstlisting} 
  ```
  
  Is a similar approach possible in Typst? Ideally the solution to this would be flexible: not only highlighting content, but also making it bold, italic, both, change the font, and so on.
  
  I have been looking around but apart from building the block manually by running over the file and manipulating each character saperately (which is far from ideal) I have not really found a good approach myself.

**Comment by xkevio** [comment]
As `raw` blocks operate on pure text (i.e. strings), it would be a bit difficult to apply formatting without weird show-rules and workarounds.
An easier way is to just create a text block with the monospace font you want and use the usual typst formatting functions. If you want, you can wrap that in a function. The only downside is of course that it won't preserve linebreaks and spaces in the same way and you have to add those manually.

```typ
#text(font: "Fira Code")[
  AA BB #text(red)[CC] DD\
  0F #highlight[67] 3E *FF*
]
```

![image](https://github.com/typst/typst/assets/13004777/ac4674ad-1424-4364-a575-9b36260b064c)


Actually, it *is* quite easy to use show-rules here but those will match on *every* occurrence of a whatever it is you're matching. But besides that, it can be quite clean and you can even limit it to only codeblocks with your own language, called `hex` for example:
```typ
#show raw.where(lang: "hex"): r => {
  show "00": set text(gray)
  show "67": highlight.with(fill: yellow)
  show "FF": strong
  
  r
}

```hex
AA BB CC DD
0F 67 3E FF
CC 00 00 00
`​`​`
```

![image](https://github.com/typst/typst/assets/13004777/6a9067da-a181-454d-8230-56ca2579e27c)

**Reply by xbmbmx** [reply]
This second option is very close to what I am trying to achieve really, and I would be fine with creatign a sort of "companion file" or rule for each code file that defines the show rules for the `raw` content manipulation.

However, the fact that it matches ALL occurances is currently a dealbreaker. I really wished there was a sort of "inline syntax" like offered by lstlisting.

I'll keep looking, if there are more recommendations please shoot. Perhaps I'll take a peek at maybe writing a plugin to handle something like this myself.

**Reply by eltos** [reply]
With a small adjustment to the show rules you can limit it to certain occurrences marked with e.g. `@`:

````typ
#show raw.where(lang: "hex"): r => {
  show "00": set text(gray)
  show "@67": it => highlight(fill: yellow, it.text.slice(1))
  show "FF": strong
  
  r
}

```hex
AA BB CC DD
0F @67 3E FF
CC 67 00 00
```

````

![test.png](https://github.com/user-attachments/assets/d863c86f-af53-43df-a078-02b093de9624)

**Comment by 8LWXpg** [comment]
Another way is to write your `raw` text line by line.
```typst
`AA BB `#text(red)[`CC `]`DD
0F `#highlight[`67`]` 3E `*`FF`*
```

![raw](https://github.com/typst/typst/assets/105704427/e975bd72-031b-45fb-a28b-09285eb893b7)

**Comment by alexantoniadis** [comment]
Hello everyone !
I hope I'll still get some help on this old discussion.

I'm trying to do the same thing. I have some big blocks of code that I want to present using the raw format and the language, but I also want to highlight some important parts of the code.

If I don't use any language when declaring the raw block, it's working well:
```
` ` `
dfir@dfir-mbp ~ % #text_bold[hello]
` ` `
```

<img width="363" alt="Capture d’écran 2024-10-28 à 19 04 48" src="https://github.com/user-attachments/assets/50d8c1b1-3fd1-400a-85c5-43d9edb54a74">

But when I add the language details in the raw block, the language formatting is erasing my custom one. 
```
` ` `bash
dfir@dfir-mbp ~ % #text_bold[hello]
` ` `
```

<img width="519" alt="Capture d’écran 2024-10-28 à 19 04 55" src="https://github.com/user-attachments/assets/f7b33636-b7e6-438e-8201-850aa7861cf7">

Is there a way to add some format rules that will be applied on an already formatted raw block?

I'm using a function to do this: 
```typst
//------------------------------------------------------------
// Bold Function
//------------------------------------------------------------

#let text_bold(body) = {text( weight: "bold", [#body])}

//------------------------------------------------------------
// Raw converter Function
//------------------------------------------------------------

#let raw_converter(body) = {
  //----------------------------------------------------------
  // Find all matches with function #text_bold
  //----------------------------------------------------------
 
  let matches_text_bold = body.text.matches(regex("#text_bold\[[^]]+\]{1,10}"))

  //----------------------------------------------------------
  // Adding the rules
  //----------------------------------------------------------
  show raw.where(): show_rule => {
    //--------------------------------------------------------
    // Set the #text_bold rule
    //--------------------------------------------------------
    for it in range(matches_text_bold.len()) {
      // Creating the value inside the #text_bold function
      let text_value = matches_text_bold.at(it).text.replace("#text_bold[","").replace("]","")
      show_rule = {
        // Remove the function to only keep the text inside
        show matches_text_bold.at(it).text: text_bold(text_value)
        show_rule
      }
    }
    show_rule
  }
  
  body
}

#let terminalbox(
  title: "",
  body_box,
) = {
      [... Creating the header of the Box ...]
      //------------------------------------------------------
      //  Body of the Box
      //------------------------------------------------------
      rect(
        width: 100%, 
        radius: (bottom-right: 6pt, bottom-left: 6pt),
        fill: body_color
        inset: (y: 8pt, x: 5pt),
        stroke: (top: 0.8pt + line_color)
        text(
          box_text_size, 
          raw_converter(body_box),
          fill: box_text_color, 
          font: box_text_type,
        ),
      ),
    )
  )
}

#terminalbox(title-bash: "Terminal",
` ` `bash
dfir@dfir-mbp ~ % #text_bold[hello]
` ` `)
```

**Reply by Marcono1234** [reply]
For future reference, it looks like you also posted this question on the forum and got an answer there: https://forum.typst.app/t/how-to-add-formatting-rules-to-an-already-formatted-raw-block/1853

(though would be good if you had added this cross-link in the first place; otherwise people repeatedly invest time trying to answer your question; and similarly others looking for an answer will not notice the answer on the other question)
